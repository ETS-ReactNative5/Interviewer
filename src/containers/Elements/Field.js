/* eslint-disable jsx-a11y/label-has-for */

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Field as ReduxFormField } from 'redux-form';
import { map, toPairs } from 'lodash';
import {
  TextInput,
  RadioGroup,
  ToggleGroup,
} from 'network-canvas-ui';

import validations from '../../utils/Validations';

import { withOptionsFromSelector } from '../../behaviours';

/**
  * Returns the named field compontent, if no matching one is found
  or else it just returns a text input
  * @param {object} field The properties handed down from the protocol form
  */

export const makeRenderInput = (type) => {
  const renderInput = (field) => {
    const {
      input,
      meta,
      label,
      options,
      optionsSelector,
      isNumericOnly,
      toggleComponent,
      autoFocus,
    } = field;

    let InputComponent = TextInput;

    let inputProps = {
      name: input.name,
      value: input.value,
      errorText: meta.invalid && meta.touched && meta.error,
      label,
      autoFocus,
      isNumericOnly,
    };

    if (type === 'RadioGroup') {
      InputComponent = RadioGroup;
      inputProps = {
        ...inputProps,
        options,
        onRadioClick: input.onChange,
      };
    }

    if (type === 'CheckboxGroup') {
      const { colors } = field;
      InputComponent = ToggleGroup;
      inputProps = {
        ...inputProps,
        toggleComponent,
        options,
        colors,
        onOptionClick: (e, checked, optionVal) => input.onChange({
          ...input.value,
          [optionVal]: checked,
        }),
      };
    }

    if (optionsSelector) {
      InputComponent = withOptionsFromSelector(InputComponent, optionsSelector);
    }

    return <InputComponent {...inputProps} {...input} />;
  };

  return renderInput;
};

/**
* Returns the named validation function, if no matching one is found it returns a validation
* which will always fail.
* @param {string} validation The name of the validation function to return.
  */
const getValidation = validation =>
  map(
    toPairs(validation),
    ([type, options]) => (
      Object.hasOwnProperty.call(validations, type) ? validations[type](options) : () => (`Validation "${type}" not found`)
    ),
  );

/**
  * Renders a redux-form field in the style of our app.
  * @param {string} label Presentational label
  * @param {string} name Property name
  * @param {string} component Field component
  * @param {string} placeholder Presentational placeholder text
  * @param {object} validation Validation methods
  */
class Field extends PureComponent {
  constructor(props) {
    super(props);
    this.component = (typeof props.component === 'string') ? makeRenderInput(props.component) : props.component;
  }

  render() {
    const { label, name, component, validation, ...rest } = this.props;

    const validate = getValidation(validation);

    return (
      <ReduxFormField
        name={name}
        label={label}
        component={this.component}
        validate={validate}
        {...rest}
      />
    );
  }
}

Field.propTypes = {
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  validation: PropTypes.object,
};

Field.defaultProps = {
  validation: {},
};

export default Field;
