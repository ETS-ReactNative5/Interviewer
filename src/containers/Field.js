import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Field as ReduxFormField } from 'redux-form';
import { map, get, toPairs } from 'lodash';
import {
  Checkbox,
  CheckboxGroup,
  ToggleButtonGroup,
  ToggleButton,
  RadioGroup,
  Text,
  Number as NumberField,
  Toggle,
  LikertScale,
  VisualAnalogScale,
  DatePicker,
} from '../ui/components/Fields';

import validations from '../utils/Validations';
import { FormComponent } from '../protocol-consts';

/*
  * Returns the named field compontent, if no matching one is found
  or else it just returns a text input
  * @param {object} field The properties handed down from the protocol form
  */

const fieldTypes = {
  [FormComponent.Checkbox]: Checkbox,
  [FormComponent.CheckboxGroup]: CheckboxGroup,
  [FormComponent.RadioGroup]: RadioGroup,
  [FormComponent.Text]: Text,
  [FormComponent.Number]: NumberField,
  [FormComponent.ToggleButton]: ToggleButton,
  [FormComponent.Toggle]: Toggle,
  [FormComponent.LikertScale]: LikertScale,
  [FormComponent.VisualAnalogScale]: VisualAnalogScale,
  [FormComponent.ToggleButtonGroup]: ToggleButtonGroup,
  [FormComponent.DatePicker]: DatePicker,
};

const ComponentTypeNotFound = componentType =>
  () => (<div>Input component &quot;{componentType}&quot; not found.</div>);

export const getInputComponent = (componentType = 'Text') =>
  get(
    fieldTypes,
    componentType,
    ComponentTypeNotFound(componentType),
  );

/**
* Returns the named validation function, if no matching one is found it returns a validation
* which will always fail.
* @param {string} validation The name of the validation function to return.
  */
const getValidation = validation =>
  map(
    toPairs(validation),
    ([type, options]) => (
      Object.hasOwnProperty.call(validations, type) ? validations[type](options) : () => (`Validation "${type}" not found`)
    ),
  );

/**
  * Renders a redux-form field in the style of our app.
  * @param {string} label Presentational label
  * @param {string} name Property name
  * @param {string} component Field component
  * @param {string} placeholder Presentational placeholder text
  * @param {object} validation Validation methods
  */
class Field extends PureComponent {
  constructor(props) {
    super(props);
    this.component = getInputComponent(props.component);
    this.validate = getValidation(props.validation);
  }

  render() {
    const { label, name, validation, ...rest } = this.props;
    return (
      <ReduxFormField
        {...rest}
        name={name}
        label={label}
        component={this.component}
        validate={this.validate}
      />
    );
  }
}

Field.propTypes = {
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  validation: PropTypes.object,
};

Field.defaultProps = {
  label: '',
  validation: {},
};

export default Field;
