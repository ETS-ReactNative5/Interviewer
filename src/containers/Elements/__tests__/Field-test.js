/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import Field, { makeRenderInput } from '../Field';

const attributes = {
  label: 'Name',
  name: 'name',
  component: 'Alphanumeric',
};

const validation = {
  required: true,
  minLength: 2,
};

const reduxFormFieldProperties = { input: { name: 'foo', value: '' }, meta: { invalid: false } };

describe('makeRenderInput()', () => {
  it('should return renderable component', () => {
    const Input = makeRenderInput('Alphanumeric');

    const subject = shallow((
      <Input {...reduxFormFieldProperties} />
    ));

    expect(subject).toMatchSnapshot();
  });
});

describe('<Field />', () => {
  it('should render', () => {
    const subject = shallow((
      <Field {...attributes} />
    ));

    expect(subject).toMatchSnapshot();
  });

  it('Loads validations from the register', () => {
    const field = shallow(<Field {...attributes} validation={validation} />);

    const validate = field.find('Field').prop('validate');

    expect(validate.length).toBe(2);
  });

  it('Loads options with optionsSelector');
});
