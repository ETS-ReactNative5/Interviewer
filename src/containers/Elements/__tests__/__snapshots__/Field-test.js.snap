// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field /> should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Field
    component={[Function]}
    label="Name"
    name="name"
    validate={Array []}
/>,
  "nodes": Array [
    <Field
      component={[Function]}
      label="Name"
      name="name"
      validate={Array []}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <Field
        component="Alphanumeric"
        label="Name"
        name="name"
        validation={Object {}}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Field {
        "_reactInternalInstance": [Circular],
        "component": [Function],
        "context": Object {},
        "props": Object {
          "component": "Alphanumeric",
          "label": "Name",
          "name": "name",
          "validation": Object {},
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Field
          component={[Function]}
          label="Name"
          name="name"
          validate={Array []}
/>,
        "_debugID": 4,
        "_renderedOutput": <Field
          component={[Function]}
          label="Name"
          name="name"
          validate={Array []}
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Field
    component="Alphanumeric"
    label="Name"
    name="name"
    validation={Object {}}
/>,
}
`;

exports[`makeRenderInput() should return renderable component 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <TextInput
    autoFocus={undefined}
    errorText={false}
    isNumericOnly={undefined}
    label={undefined}
    name="foo"
    value=""
/>,
  "nodes": Array [
    <TextInput
      autoFocus={undefined}
      errorText={false}
      isNumericOnly={undefined}
      label={undefined}
      name="foo"
      value=""
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <renderInput
        input={
                Object {
                        "name": "foo",
                        "value": "",
                      }
        }
        meta={
                Object {
                        "invalid": false,
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "input": Object {
            "name": "foo",
            "value": "",
          },
          "meta": Object {
            "invalid": false,
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <TextInput
          autoFocus={undefined}
          errorText={false}
          isNumericOnly={undefined}
          label={undefined}
          name="foo"
          value=""
/>,
        "_debugID": 2,
        "_renderedOutput": <TextInput
          autoFocus={undefined}
          errorText={false}
          isNumericOnly={undefined}
          label={undefined}
          name="foo"
          value=""
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <renderInput
    input={
        Object {
            "name": "foo",
            "value": "",
          }
    }
    meta={
        Object {
            "invalid": false,
          }
    }
/>,
}
`;
